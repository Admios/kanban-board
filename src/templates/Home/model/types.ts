export interface Todo {
  id: string;
  stateId: string;
  text: string;
  position: number;
  owner: string;
}

export interface State {
  id: string;
  name: string;
  position: number;
  color: string;
  owner: string;
}
