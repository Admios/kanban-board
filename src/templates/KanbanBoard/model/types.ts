export interface Task {
  id: string;
  stateId: string;
  text: string;
  position: number;
}

export interface State {
  id: string;
  boardId: string;
  name: string;
  position: number;
  color: string;
}

export interface Board {
  id: string;
  name: string;
  owner: string;
}
