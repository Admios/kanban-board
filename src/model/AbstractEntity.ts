/**
 * Entities represent how the data is stored in the database.
 */
export interface AbstractEntity extends Record<string, any> {}
