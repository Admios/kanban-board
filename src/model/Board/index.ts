export { BoardRepository } from "./BoardRepository";
export { type BoardDTO } from "./BoardDTO";
