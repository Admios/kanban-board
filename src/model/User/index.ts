export { UserRepository } from "./UserRepository";
export { type UserDTO } from "./UserDTO";
