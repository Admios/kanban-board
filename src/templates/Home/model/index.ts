import { StateDTO } from "@/model/State";
import { TaskDTO } from "@/model/Task";
import { UserDTO } from "@/model/User";
import { Immutable, produce } from "immer";
import { StateCreator, create } from "zustand";
import { State, Task } from "./types";

export type { State, Task } from "./types";

type HomeState = Immutable<{
  tasks: Record<string, Task[]>;
  states: Record<string, State>;
  user?: UserDTO;
}>;

interface HomeActions {
  initialize(
    initialTasks: TaskDTO[],
    initialStates: StateDTO[],
    initialUser?: UserDTO,
  ): void;
  addTask(newTask: Task): Task;
  moveTask(
    newTask: Task,
    fromStateId: string,
    toStateId: string,
    position: number,
  ): void;
  addState(newState: State): State;
  editTask(id: string, updatedValues: Partial<Task>): void;
  deleteTask: (id: string) => void;
}

const stateCreator: StateCreator<HomeState & HomeActions> = (set, get) => ({
  tasks: {},
  states: {},

  initialize(initialTasks, initialStates, initialUser) {
    const statesMap: Record<string, State> = {};
    const tasksMap: Record<string, Task[]> = {};

    initialStates.forEach((backendState) => {
      const result: State = {
        ...backendState,
        id: backendState.id,
        position: backendState.position,
      };
      statesMap[result.id] = result;
    });

    initialTasks.forEach((backendTask) => {
      const { stateId } = backendTask;
      if (!tasksMap[stateId]) {
        tasksMap[stateId] = [];
      }
      tasksMap[stateId][backendTask.position] = {
        ...backendTask,
        stateId,
        id: backendTask.id,
        position: backendTask.position,
      };
    });

    set({ tasks: tasksMap, states: statesMap, user: initialUser });
  },

  addTask: (newTask) => {
    const result = { ...newTask };
    const tasks = produce(get().tasks, (draft) => {
      if (!draft[newTask.stateId]) {
        draft[newTask.stateId] = [];
      }

      draft[newTask.stateId].push(result);
    });

    set({ tasks });
    return result;
  },

  moveTask: (newTask, fromStateId, toStateId, position) => {
    const tasks = produce(get().tasks, (draftState) => {
      const sourceState = (draftState[fromStateId] ?? []).filter(
        (task) => task.id !== newTask.id,
      );

      let destinationState = draftState[toStateId] ?? [];
      if (fromStateId === toStateId) {
        destinationState = sourceState;
      }

      if (position <= destinationState.length) {
        destinationState.splice(position, 0, newTask);
      } else {
        destinationState.push(newTask);
      }

      draftState[fromStateId] = sourceState;
      draftState[toStateId] = destinationState;
    });

    set({ tasks });
  },

  addState: (newState) => {
    const currentModel = get();
    const tasks = produce(currentModel.tasks, (draft) => {
      draft[newState.id] = [];
    });
    const states = produce(currentModel.states, (draft) => {
      const position = Object.keys(currentModel.states).length;
      draft[newState.id] = { ...newState, position };
    });

    set({ states, tasks });
    return states[newState.id];
  },

  editTask: (taskId, updatedValues) => {
    const currentState = get();
    const tasks = produce(currentState.tasks, (draft) => {
      for (const stateId in draft) {
        const stateTasks = draft[stateId];
        const taskIndex = stateTasks.findIndex((task) => task.id === taskId);
        if (taskIndex > -1) {
          stateTasks[taskIndex] = {
            ...stateTasks[taskIndex],
            ...updatedValues,
          };
          break;
        }
      }
    });

    set({ tasks });
  },

  deleteTask: (taskId) => {
    const currentState = get();
    const tasks = produce(currentState.tasks, (draft) => {
      for (const stateId in draft) {
        const stateTasks = draft[stateId];
        const taskIndex = stateTasks.findIndex((task) => task?.id === taskId);
        if (taskIndex > -1) {
          stateTasks.splice(taskIndex, 1);
          break;
        }
      }
    });

    set({ tasks });
  },
});

export const useZustand = create(stateCreator);
