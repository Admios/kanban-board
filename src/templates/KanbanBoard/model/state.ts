import { StateDTO } from "@/model/State";
import { TaskDTO } from "@/model/Task";
import { UserDTO } from "@/model/User";
import { Immutable } from "immer";
import { State, Task } from "./types";

export type KanbanState = Immutable<{
  /**
   * The order of the states in the board.
   */
  statesOrder: string[];

  /**
   * The order of the tasks in each state.
   */
  tasksOrder: Record<string, string[]>;

  /**
   * Key: taskId. Value: Task.
   */
  tasks: Record<string, Task>;

  /**
   * Key: stateId. Value: State.
   */
  states: Record<string, State>;

  /**
   * The user that is currently logged in
   */
  user?: UserDTO;
}>;

export interface KanbanActions {
  setUser(user?: UserDTO): void;
  initialize(initialTasks: TaskDTO[], initialStates: StateDTO[]): void;
  addTask(newTask: Task): Task;
  moveTask(
    newTask: Task,
    fromStateId: string,
    toStateId: string,
    position: number,
  ): void;
  addState(newState: State): State;
  editTask(id: string, updatedValues: Partial<Task>): void;
  deleteTask: (id: string) => void;

  clear(): void;
}
