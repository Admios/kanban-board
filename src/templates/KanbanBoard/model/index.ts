import { produce } from "immer";
import { StateCreator, create } from "zustand";
import { State, Task } from "./types";
import { KanbanState, KanbanActions } from "./state";

export type { State, Task } from "./types";

const stateCreator: StateCreator<KanbanState & KanbanActions> = (set, get) => ({
  statesOrder: [],
  tasksOrder: {},
  tasks: {},
  states: {},

  initialize(initialTasks, initialStates) {
    const statesOrderSparseArray: string[] = [];
    const tasksOrder: Record<string, string[]> = {};
    const states: Record<string, State> = {};
    const tasks: Record<string, Task> = {};

    initialTasks.forEach((backendTask) => {
      tasks[backendTask.id] = backendTask;
      if (!tasksOrder[backendTask.stateId]) {
        tasksOrder[backendTask.stateId] = [];
      }
      // Warning: saved as sparse array! Compression is done later
      tasksOrder[backendTask.stateId][backendTask.position] = backendTask.id;
    });

    initialStates.forEach((backendState) => {
      states[backendState.id] = backendState;
      // Warning: saved as sparse array! compressed later
      statesOrderSparseArray[backendState.position] = backendState.id;
      // Compress sparse array
      tasksOrder[backendState.id] = (tasksOrder[backendState.id] ?? []).filter(
        (id) => !!id,
      );
    });

    // Compress sparse array
    const statesOrder = statesOrderSparseArray.filter((id) => !!id);

    set({
      statesOrder,
      tasksOrder,
      tasks,
      states,
    });
  },

  setUser(user) {
    set({ user });
  },

  addTask: (newTask) => {
    const tasks = produce(get().tasks, (draft) => {
      draft[newTask.stateId] = newTask;
    });

    const tasksOrder = produce(get().tasksOrder, (draft) => {
      if (!draft[newTask.stateId]) {
        draft[newTask.stateId] = [];
      }
      draft[newTask.stateId].push(newTask.id);
    });

    set({ tasks, tasksOrder });
    return newTask;
  },

  moveTask: (newTask, fromStateId, toStateId, position) => {
    const newTasksOrder = produce(get().tasksOrder, (draftState) => {
      const sourceState = (draftState[fromStateId] ?? []).filter(
        (task) => task !== newTask.id,
      );

      let destinationState = draftState[toStateId] ?? [];
      if (fromStateId === toStateId) {
        destinationState = sourceState;
      }

      if (position <= destinationState.length) {
        destinationState.splice(position, 0, newTask.id);
      } else {
        destinationState.push(newTask.id);
      }

      draftState[fromStateId] = sourceState;
      draftState[toStateId] = destinationState;
    });

    set({ tasksOrder: newTasksOrder });
  },

  addState: (newState) => {
    const tasksOrder = produce(get().tasksOrder, (draft) => {
      draft[newState.id] = [];
    });

    const statesOrder = produce(get().statesOrder, (draft) => {
      draft.push(newState.id);
    });

    const states = produce(get().states, (draft) => {
      draft[newState.id] = newState;
    });

    set({ states, tasksOrder, statesOrder });
    return states[newState.id];
  },

  editTask: (taskId, updatedValues) => {
    const tasks = produce(get().tasks, (draft) => {
      draft[taskId] = { ...draft[taskId], ...updatedValues };
    });

    set({ tasks });
  },

  deleteTask: (taskId) => {
    const { stateId } = get().tasks[taskId];
    const tasksOrder = produce(get().tasksOrder, (draft) => {
      draft[stateId] = draft[stateId].filter((id) => id !== taskId);
    });

    const tasks = produce(get().tasks, (draft) => {
      delete draft[taskId];
    });

    set({ tasks, tasksOrder });
  },
});

export const useZustand = create(stateCreator);
