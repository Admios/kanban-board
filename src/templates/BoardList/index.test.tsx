import { render, screen, within } from "@testing-library/react";
import { BoardList } from "./index";
import { UserDTO } from "@/model/User";
import { BoardDTO } from "@/model/Board";

const mockUser: UserDTO = {
  email: "John Doe",
};

const mockBoards: BoardDTO[] = [
  {
    id: "1",
    name: "Test Board",
    owner: "owner@test.com",
  },
  {
    id: "2",
    name: "Another Board",
    owner: "owner@test.com",
  },
];

describe("BoardList", () => {
  it("renders the navbar with the user's name", () => {
    const { container } = render(
      <BoardList user={mockUser} boards={mockBoards} />,
    );
    expect(container).toMatchSnapshot("Normal Layout");
  });

  it("renders the board links", () => {
    render(<BoardList user={mockUser} boards={mockBoards} />);
    const boardLinks = within(screen.getByTitle("boards-list")).getAllByRole(
      "link",
      { name: /board/i },
    );
    expect(boardLinks).toHaveLength(mockBoards.length);
  });

  it("renders an empty layout when there are no boards", () => {
    const { container } = render(<BoardList user={mockUser} boards={[]} />);
    expect(container).toMatchSnapshot("Empty Layout");
  });
});
