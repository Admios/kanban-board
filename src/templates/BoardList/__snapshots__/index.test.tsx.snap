// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardList renders the navbar with the user's name 1`] = `
<div>
  <main
    class="main"
  >
    <nav
      aria-label="main navigation"
      class="navbar"
      role="navigation"
    >
      <div
        class="navbar-brand"
      >
        <a
          class="navbar-item"
          href="/"
        >
          <img
            alt="Logo"
            data-nimg="1"
            decoding="async"
            height="28"
            loading="lazy"
            src="/admios-logo.svg"
            style="color: transparent;"
            width="112"
          />
        </a>
        <a
          aria-expanded="false"
          aria-label="menu"
          class="navbar-burger"
          data-target="main-navbar-inner"
          role="button"
        >
          <span
            aria-hidden="true"
          />
          <span
            aria-hidden="true"
          />
          <span
            aria-hidden="true"
          />
        </a>
      </div>
      <div
        class="navbar-menu"
        id="main-navbar-inner"
      >
        <div
          class="navbar-start"
        >
          <a
            class="navbar-item"
            href="/"
          >
            My Boards
          </a>
        </div>
        <div
          class="navbar-end"
        >
          <div
            class="navbar-item"
          >
            Username: 
            John Doe
          </div>
          <a
            class="navbar-item"
          >
            Logout
          </a>
        </div>
      </div>
    </nav>
    <section
      class="container"
      title="boards-list"
    >
      <nav
        class="panel is-primary"
      >
        <p
          class="panel-heading"
        >
          Your Boards
        </p>
        <a
          class="panel-block is-active"
          href="/b/1"
        >
          <span
            class="panel-icon"
          >
            <svg
              aria-hidden="true"
              data-slot="icon"
              fill="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18.375 2.25c-1.035 0-1.875.84-1.875 1.875v15.75c0 1.035.84 1.875 1.875 1.875h.75c1.035 0 1.875-.84 1.875-1.875V4.125c0-1.036-.84-1.875-1.875-1.875h-.75ZM9.75 8.625c0-1.036.84-1.875 1.875-1.875h.75c1.036 0 1.875.84 1.875 1.875v11.25c0 1.035-.84 1.875-1.875 1.875h-.75a1.875 1.875 0 0 1-1.875-1.875V8.625ZM3 13.125c0-1.036.84-1.875 1.875-1.875h.75c1.036 0 1.875.84 1.875 1.875v6.75c0 1.035-.84 1.875-1.875 1.875h-.75A1.875 1.875 0 0 1 3 19.875v-6.75Z"
              />
            </svg>
          </span>
          Test Board
        </a>
        <a
          class="panel-block is-active"
          href="/b/2"
        >
          <span
            class="panel-icon"
          >
            <svg
              aria-hidden="true"
              data-slot="icon"
              fill="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18.375 2.25c-1.035 0-1.875.84-1.875 1.875v15.75c0 1.035.84 1.875 1.875 1.875h.75c1.035 0 1.875-.84 1.875-1.875V4.125c0-1.036-.84-1.875-1.875-1.875h-.75ZM9.75 8.625c0-1.036.84-1.875 1.875-1.875h.75c1.036 0 1.875.84 1.875 1.875v11.25c0 1.035-.84 1.875-1.875 1.875h-.75a1.875 1.875 0 0 1-1.875-1.875V8.625ZM3 13.125c0-1.036.84-1.875 1.875-1.875h.75c1.036 0 1.875.84 1.875 1.875v6.75c0 1.035-.84 1.875-1.875 1.875h-.75A1.875 1.875 0 0 1 3 19.875v-6.75Z"
              />
            </svg>
          </span>
          Another Board
        </a>
      </nav>
    </section>
  </main>
</div>
`;
